### PR TITLE
Added piplines/subpipelines functionality

### PR DESCRIPTION
**Pull request by AlanAboudib**
## Description

Created pipelines/subpipelines functionality. Here is an example usage:

```
from syfertext.pipeline import SimpleTagger
from syft.generic.string import String

# Create a simple tagger, to tag some stop words
tagger = SimpleTagger(attribute= 'is_stop',
                      lookups=['I', 'myself'],
                      case_sensitive= False,
                      tag=True,
                      default_tag=False)

# Create a text as a PySyft string
text = String("I like to do things myself")

# Add the tagger to the pipeline
nlp.add_pipe(component=tagger, name='my tagger')

# Tokenize and then tag stop words
doc = nlp(text)

# Check out whether the tokens were correctly tagged
for token in doc:
    print(token.text.ljust(7), "|", token._.is_stop)

# output:
# I       | True
# like    | False
# to      | False
# do      | False
# things  | False
# myself  | True
```

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

- `tests/distributed_mode/test_pipeline.py`
- All tests that use the tokenizer, test the pipelines/subpipelines indirectly.

## Checklist:

* [x] I did follow the [contribution guidelines](https://github.com/OpenMined/PySyft/blob/master/CONTRIBUTING.md)
* [x] I have commented my code following [Google style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).
* [x] I have added tests for my changes

### CODE REVIEW

**Comment by sachin-101** [comment]
@AlanAboudib If we pass last = false, in add_pipe
https://github.com/OpenMined/SyferText/blob/c0fed6771535fefacb02aa3fdaf1b5c9290a5b98/syfertext/language.py#L244
Still this logic false back to adding the component at the end. Is it desirable ?

**Comment by AlanAboudib** [comment]
@sachin-101 

That is a good catch. I agree it is not very elegant. But it is subtle. How would you suggest to  solve this?

I think one way would be to make the default value for `last` as `None` not `False` then 

`if last or (last is None and not any ([before, after, first]))`

**Comment by sachin-101** [comment]
So, according to your logic, if last is set to either True or False, then `last is None and not any([before, after, first])` is always False. So it depends only on 'last', which boils down to `if last`.
But if last is not passed and last = None is default, 
1.  before/after/first is also not passed, then the component is added at the end, which is last = True when last is None.
2. If any of before/after/first is set to True, then the position will be set accordingly, since `(last is None and not any ([before, after, first]))` will be False.

Also we are already checking whether only one is set using your sum < 2 logic, so these 3 cases are the only possibilities.
So, setting last = True by default and `if last` also does the same as 
setting last = None and `if last or (last is None and not any ([before, after, first]))`

**Comment by sachin-101** [comment]
@AlanAboudib It boils down to what we wish to allow the user to do. I suggest, if he is setting `last=False` explicitly but not setting any other `position = True`, then we should throw an error rather than adding the component at last.
This we can do by `assert(
            sum([bool(before), bool(after), bool(first), bool(last)]) == 1
        )`.

**Comment by AlanAboudib** [comment]
@sachin-101 I think if we change nothing in the code, it would be ok. Because if you set `last == False` you will get an error anyway.

**Comment by AlanAboudib** [comment]
Oh excuse me, my last comment is wrong. 

What you suggest:

`assert( sum([bool(before), bool(after), bool(first), bool(last)]) == 1`

Seems a good choice for me.

Then we no more need to change anything in the code

**Comment by AlanAboudib** [comment]
Would you like to add some tests for `add_pipe()`?

**Comment by sachin-101** [comment]
Ok! I will do it.

File: syfertext/language.py [hunk 1]
```diff
@@ -84,55 +70,341 @@ def __init__(
...
+
+        # Delete the pipe using its index
+        pipe = self.pipeline_template.pop(pipe_index)
+
+        # Parse the pipeline template again
+        # to create the subpipeline templates
+        self._parse_pipeline_template()
+
+        # Reset the pipeline.
+        self._reset_pipeline()
```
Comment (bicycleman15) [inline syfertext/language.py:300]: @AlanAboudib Do we need to call `_parse_pipeline_template` ? I think its being already done in `_reset_pipeline`. So Its being done twice currently.

Comment (AlanAboudib) [reply syfertext/language.py:300]: @bicycleman15 Good catch. I have changed `_reset_pipeline()` to include `_parse_pipeline_templte()` and forgot to remove it here. Do not hesitate to create a PR for this

**Pull request merged**